add aten::empty_strided.SymInt

bdhirsh opened the pull request:
Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81302
* #81301
* __->__ #81238

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

bdhirsh committed:
add aten::empty_strided.SymInt

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/EmptyTensor.cpp:
```diff
@@ -98,6 +98,37 @@ size_t computeStorageNbytes(
+    // TODO: this should be (sizes[i] - 1), but that breaks, figure out why.
+    auto tmp1 = sizes[i];
+    auto tmp2 = strides[i];
+    SymInt next_size = tmp1 * tmp2;
```
@Chillee this line is giving me problems locally - do you have any idea of what's going on? Error:
```
RuntimeError: NotImplementedError: argument of type: <class 'torch.fx.experimental.symbolic_shapes.PySymInt'>
```

bdhirsh replied:
Still digging, but I have a feeling it has to do with the way `stride()` is meant to be implemented in `FakeTensor` (maybe I did it wrong?) It's in the previous PR here: https://github.com/pytorch/pytorch/pull/81237/files#diff-39e82af71afdadbc56a4ecf552ec668ddcda794f8ea3ec41edaef23456fd56e9R474

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels




[ghstack-poisoned]

ezyang commented on aten/src/ATen/FunctionalStorageImpl.cpp:
```diff
@@ -89,10 +89,19 @@ bool Alias::apply_updates() {
   return any_updates;
 }
 
+// Should we just add a sym_numel() so we don't need to do this?
```
yes we should

ezyang commented on aten/src/ATen/EmptyTensor.cpp:
```diff
@@ -98,6 +98,27 @@ size_t computeStorageNbytes(
 
+// not including mobile-only macros in this function,
+// since mobile shouldn't be using symints.
+SymInt computeStorageNbytesSymInt(
```
it would be nice to get `aten::empty` using this helper too

bdhirsh replied:
Yep... it looks like it might require some larger changes to `PySymInt` et.al, so I figured I wouldn't bite it off in this PR. LMK what you think though.


It looks like the way that `SymInt::operator+` (and the others) are implemented today, if one SymInt is symbolic and the other isn't, it'll effectively promote the non-symint to a real `SymbolicIntNode` object. Which I don't think is what we want here? If `at::empty()` is meant to construct a contiguous tensor, it seems weird to introduce a bunch of extra sym stride nodes in the expression to compute the number of bytes.

ezyang replied:
Why isn't it what you want? If a size is symbolic you are going to need a symbolic storage size too.

bdhirsh replied:
What I mean is, the symbolic expression computed by `at::empty` today looks something like:
```
storage_size = size * size2 * size3 * ...
```

And the symbolic expression you'll get from running `computeStorageNbytesSymInt` looks a lot more complicated, since it involves symbolic strides:
```
(storage_offset) + 1 + ([size1 - 1] * stride1) + ([size2 - 1] * stride2) + ([size3 - 1] * stride3) + ...
```

Is the idea that this is ok, because it should be possible to simplify that expression back down to the contiguous version? Since `strideN` variables above would actually be written in terms of sizes.

ezyang replied:
Oh, this is fair. I suppose a simplified path when the strides are contiguous is a reasonable idea.

ezyang replied:
In principle if you guard the strides enough (as they are contiguous), yes they should be equivalent, but we'd have to check if sympy can actualy figure it out.

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/MetaTensor.cpp:
```diff
@@ -53,5 +53,27 @@ Tensor empty_strided_meta(
+  c10::optional<bool> pin_memory_opt
+) {
+  return at::detail::empty_strided_symint_meta(
+      size, stride, dtype_opt, layout_opt, device_opt, pin_memory_opt);
```
??

bdhirsh replied:
This is mostly cargo-culted from how `empty_symint_meta()` is implemented. It looks like `at::native::empty_symint_meta` is responsible for checking if the sizes are actually symbolic, and shunting to either the non-symbolic kernel (`at::native::empty_meta`), or the symbolic kernel (`at::detail::empty_symint_meta`)

ezyang replied:
No what I mean is, you have a bunch of dead code below the return

bdhirsh replied:
... oh welp haha

ezyang commented on aten/src/ATen/native/mps/TensorFactory.cpp:
```diff
@@ -101,6 +101,25 @@ Tensor empty_strided_mps(
+      device_opt,
+      pin_memory_opt);
+  resize_impl_mps_(t.unsafeGetTensorImpl(), c10::asIntArrayRefSlow(size), c10::asIntArrayRefSlow(stride));
+  return t;
```
I expected this to call `empty_strided_mps` directly

bdhirsh replied:
Yep that's fair. This mostly just replicates the logic of `empty_strided_mps` (uglier)

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2077,6 +2077,14 @@
+    CUDA: empty_strided_symint_cuda
+    MPS: empty_strided_symint_mps
+    Meta: empty_strided_symint_meta
+    QuantizedCPU, QuantizedCUDA: empty_strided_symint_unknown_quantized
```
Now that https://github.com/pytorch/pytorch/pull/82189 is in it's not necessary to dupe the dispatch table here. @Krovatkin

bdhirsh replied:
hmmm why is that? It seems like in master today, we still dupe the dispatch table for `empty.memory_format` and `empty.SymInt`.

ezyang replied:
yeah we haven't fixed it yet

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot help

bdhirsh commented:
@pytorchbot --help

bdhirsh commented:
cc @pytorch/pytorch-dev-infra, I pushed a commit to my PR, which I'd expect to re-run CI on all 4 PR's in the stack. But it looks like CI is only re-running on the top-most PR. If I'm reading the HUD page correctly, it seems like no new jobs were triggered on the HUD either: https://hud.pytorch.org/pr/81238

Just an fyi, but I'm going to force push to try to re-trigger CI

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.




[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/FunctionalStorageImpl.cpp:
```diff
@@ -89,10 +90,23 @@ bool Alias::apply_updates() {
+    // LTC hasn't implemented SymInt support yet though (torch::lazy::SymIntNodeImpl).
+    // Once it does, we should remove this check.
+    if (value.key_set().has(c10::DispatchKey::Python)) {
+      return at::detail::computeStorageNbytesSymInt(value.sym_sizes(), value.sym_strides(), value.dtype().itemsize(), value.storage_offset());
```
cc @Krovatkin I ended up doing this to fix LTC test failures. Just confirming that this seems ok to you?

Originally, I wrote this code to say "if the inner tensor is using symbolic strides, then do symint stuff".

That breaks LTC though, because none of the methods on `torch::lazy::SymIntNodeImpl` are implemented yet so it can't actually trace any symbolic shape stuff yet.

I figure that once LTC/XLA are ready to trace symbolic shapes, then we can fix this?

ezyang commented on aten/src/ATen/EmptyTensor.cpp:
```diff
@@ -163,6 +184,27 @@ TensorBase empty_strided_generic(
+      std::move(storage_impl), ks, dtype);
+  tensor.unsafeGetTensorImpl()->set_sym_sizes_and_strides(size, stride);
+  return tensor;
+}
```
@bdhirsh what do you feel about just converting these functions into templates. Reaaally trying to avoid duping them

bdhirsh replied:
Agreed that the code duping feels bad.

Are you thinking of fully converting the headers into templates, like `at::detail::empty_strided_meta<IntArrayRef>`? (removing all the duplication at the cost of the API requiring templates).

Or keeping the API keeping `empty_strided_meta` + `empty_strided_symint_meta` around, and just internally wrapping their implementation in a shared template. (which... would cut down on some duplication, but I most of the LoC add here just comes from the extra signatures being added.)

Just want to confirm because the first one probably involves updating a bunch of call-sites :p.

Also, are you suggesting to rename `set_sym_sizes_and_strides(SymIntArrayRef sizes, ...)` to `set_sizes_and_strides(SymIntArrayRef sizes, ...)`? Otherwise I'm not sure how we can really de-dup this particular function (`empty_strided_symint_generic`).

ezyang replied:
Latter.

ezyang commented on c10/core/SymInt.cpp:
```diff
@@ -43,6 +43,14 @@ SymInt SymInt::operator+(SymInt sci) const {
+  auto res = normalize_symints(*this, sci);
+  return SymInt::toSymInt(res[0]->sub(res[1]));
+}
+
```
we should land this sooner rather than later

ezyang reviewed (approved):
In the spirit of moving things along

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge

zhxchen17 commented:
Hi @bdhirsh just wondering, do we plan to land this stack recently? want to try out functionalization with dynamic shape tracing for export, so it'd be great if this could be shipped, thanks!

bdhirsh commented:
@zhxchen17 this PR is stalled a bit on landing https://github.com/pytorch/pytorch/pull/83628 (which impacts how we write symint overloads). Once that lands I'll try to push this over the line.

bdhirsh committed:
Update on "add aten::empty_strided.SymInt"

Adds `aten::empty_strided.SymInt`, which is needed to get functionalization to plumb symints properly.

Specifically, all of the functionalization kernels use `empty_strided` to generate meta tensors, that they use internally to figure out the correct sizes/strides of their outputs. I update that code (`to_meta()`) to use the new version, `empty_strided.SymInt`.


Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

bdhirsh commented:
Updated this PR again on top of the new symint codegen changes. I'm going on PTO for 1.5 weeks starting tomorrow though, so I probably won't be able to land this before then :( If someone really needs the change in before then, feel free to grab it from me.

At the very least, it still needs:
- Fixes for the failing LTC tests
- (probably) and XLA-side patch to `empty.symint`
- importing to fbcode to check for any internal breakages, since backends registering stuff to `empty.symint` will need to change their kernel signatures.

bdhirsh commented on torch/csrc/lazy/core/tensor_impl.cpp:
```diff
@@ -148,6 +157,18 @@ void LTCTensorImpl::setup_sym_sizes() const {
+    for (auto i = last_idx - 1; i >= 0; --i) {
+      sym_strides.at(i) = sym_strides.at(i+1) * sym_sizes.at(i+1);
+    }
+  }
```
@Krovatkin one question I have is - how far along is LTC symbolic shapes / is it far enough along to try to get symbolic strides working with LTC in this PR?

Also, let me know if what I put seems at all reasonable to you - I basically:
-  matched LTC's `sym_sizes_` field with a `sym_strides_` field
- for simplicity (although it didn't work out that way haha), I assumed contiguous strides.
- Doing that requires `operator*` to be implemented on LTC's `torch::lazy::SymIntNodeImpl`, so I moved it around and started trying to add a few methods to it.

Right now, the tests currently fail because `torch::lazy::SymIntNodeImpl::wrap()` currently isn't implemented, and I wasn't sure if that was in a state where it's ready to work on yet.

Let me know if you think we should just do something totally different in this PR!
